File: etamen-backend/app/Modules/Medications/Infrastructure/Models/MedicationReminder.php
<?php

namespace App\Modules\Medications\Infrastructure\Models;

use App\Models\User;
use App\Modules\Medications\Domain\Enums\MedicationFrequencyType;
use App\Modules\Medications\Domain\Enums\MedicationReminderSource;
use App\Modules\Medications\Domain\Enums\MedicationReminderStatus;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class MedicationReminder extends Model
{
    protected $fillable = [
        'patient_user_id',
        'medication_name',
        'dosage',
        'dosage_unit',
        'instructions',
        'frequency_type',
        'interval_hours',
        'start_date',
        'end_date',
        'timezone',
        'status',
        'prescribed_by',
        'notes',
        'refill_enabled',
        'refill_quantity',
        'refill_threshold',
        'refill_reminder_date',
        'source',
        'metadata',
    ];

    protected function casts(): array
    {
        return [
            'frequency_type' => MedicationFrequencyType::class,
            'interval_hours' => 'integer',
            'start_date' => 'date',
            'end_date' => 'date',
            'status' => MedicationReminderStatus::class,
            'refill_enabled' => 'boolean',
            'refill_quantity' => 'integer',
            'refill_threshold' => 'integer',
            'refill_reminder_date' => 'date',
            'source' => MedicationReminderSource::class,
            'metadata' => 'array',
        ];
    }

    public function patient(): BelongsTo
    {
        return $this->belongsTo(User::class, 'patient_user_id');
    }

    public function times(): HasMany
    {
        return $this->hasMany(MedicationReminderTime::class)->orderBy('time_of_day');
    }

    public function activeTimes(): HasMany
    {
        return $this->times()->where('is_active', true);
    }

    public function logs(): HasMany
    {
        return $this->hasMany(MedicationLog::class);
    }

    public function refillEvents(): HasMany
    {
        return $this->hasMany(MedicationRefillEvent::class);
    }

    public function notificationQueue(): HasMany
    {
        return $this->hasMany(MedicationNotificationQueue::class);
    }
}

File: etamen-flutter/lib/features/medications/domain/entities/medication_reminder.dart
import 'package:etamen_app/features/medications/domain/entities/medication_reminder_time.dart';

enum MedicationReminderStatus {
  active('active'),
  paused('paused'),
  completed('completed'),
  cancelled('cancelled'),
  unknown('unknown');

  const MedicationReminderStatus(this.wireValue);

  final String wireValue;

  static MedicationReminderStatus fromWire(Object? value) {
    final raw = value?.toString().trim().toLowerCase();
    return MedicationReminderStatus.values.firstWhere(
      (item) => item.wireValue == raw || item.name.toLowerCase() == raw,
      orElse: () => MedicationReminderStatus.unknown,
    );
  }
}

enum MedicationFrequencyType {
  onceDaily('once_daily'),
  twiceDaily('twice_daily'),
  threeTimesDaily('three_times_daily'),
  customTimes('custom_times'),
  everyXHours('every_x_hours'),
  specificDays('specific_days'),
  asNeeded('as_needed'),
  unknown('unknown');

  const MedicationFrequencyType(this.wireValue);

  final String wireValue;

  static MedicationFrequencyType fromWire(Object? value) {
    final raw = value?.toString().trim().toLowerCase();
    return MedicationFrequencyType.values.firstWhere(
      (item) => item.wireValue == raw || item.name.toLowerCase() == raw,
      orElse: () => MedicationFrequencyType.unknown,
    );
  }
}

class MedicationReminder {
  const MedicationReminder({
    required this.id,
    required this.medicationName,
    required this.frequencyType,
    required this.status,
    this.dosage,
    this.dosageUnit,
    this.instructions,
    this.intervalHours,
    this.startDate,
    this.endDate,
    this.timezone,
    this.prescribedBy,
    this.notes,
    this.refillEnabled = false,
    this.refillQuantity,
    this.refillThreshold,
    this.refillReminderDate,
    this.times = const [],
    this.disclaimer,
    this.createdAt,
    this.updatedAt,
  });

  final int id;
  final String medicationName;
  final String? dosage;
  final String? dosageUnit;
  final String? instructions;
  final MedicationFrequencyType frequencyType;
  final int? intervalHours;
  final String? startDate;
  final String? endDate;
  final String? timezone;
  final MedicationReminderStatus status;
  final String? prescribedBy;
  final String? notes;
  final bool refillEnabled;
  final int? refillQuantity;
  final int? refillThreshold;
  final String? refillReminderDate;
  final List<MedicationReminderTime> times;
  final String? disclaimer;
  final DateTime? createdAt;
  final DateTime? updatedAt;

  String get dosageText {
    final parts = [dosage, dosageUnit]
        .where((item) => item?.trim().isNotEmpty == true)
        .map((item) => item!.trim())
        .toList(growable: false);
    return parts.join(' ');
  }

  bool get canPause => status == MedicationReminderStatus.active;

  bool get canResume => status == MedicationReminderStatus.paused;

  bool get canCancel =>
      status == MedicationReminderStatus.active ||
      status == MedicationReminderStatus.paused;
}
